### DISCUSSION TITLE
Please clarify user and intall directories

### DISCUSSION THREAD

**Discussion by iandoug** [sile-typesetter/sile#1464] [Q&A] [answered]
Hi

Can the install instructions / manual maybe clarify (for installing source on Linux)

1. which user ... root or me?
2. which folder to do things from.

At the moment moment I seem to have things in too places ... /usr/local/bin/sile and the sile-0.12.4 buried subfolder where I installed from.

I had to run the install as root but some things only work when you are in the source folder. Others don't.

(as root in sile folder)  sile -e 'installPackage("fontproof");os.exit()'

"As long as you run all your SILE files from within the original installation folder, this is all you need to do." ... Nope

(as me in sile folder)
Trying sile fpTest.sil to generate a fontbook gives
SILE v0.12.4 (Lua 5.4)
! Unable to find file fpTest.silno file 'fpTest.sil' at         fpTest.sil
! Could not find file at        fpTest.sil

Thanks, Ian

**Comment by alerque** [comment]
Always download, extract, configure, and build software as your user. Then as the very last step if you are installing to a base system manually (i.e. not using the system package manager) you'll need to do the very last step as root. So something like `./configure` → `make` → `sudo make install`.

You should either build and use SILE from the source directory *without* installing it, or install it and *use the installed one*, but don't mix and match. Don't install it to your system and then try to run it from inside the source directory. Having some or all bits installed to default paths on your system and also being in the source directory will confuse everything. It *can* be done but unless you are very careful to override various environment variables you'll get a Frankenstein of mixed sources.

At the configure step, you do have the option of using the typical `--prefix` to override your system's default install location. Evidently yours is `/usr/local` — which is pretty typical for software installed outside of the system package manager. I recommend you stick with that for manual source installs.

As such you'll end up with stuff in several folders, the executable should be in `/usr/local/bin`, some header files in `/usr/local/include/libtexpdf`, some shared libraries in `/usr/local/lib/sile`, and a bunch of Lua stuff in `/usr/local/share/sile`. These are all pretty standard locations for Linux application installs (and indeed we're just using the autoconf defaults).

By the way which system are you on that doesn't have 0.12.4 packages ready to go without needing a source build?

The issue with installing the fontproof package is ... a different issue. We can look into that, but lets confirm whether you have a working SILE install or not first with a simpler self check:

```console
$ cd /tmp
$ echo '<sile>test</sile>' | sile /dev/stdin -o foo.pdf
SILE v0.12.4 (Lua 5.4)
</dev/stdin>
[1]
```

You should get a PDF file at /tmp/foo.pdf. Do you?

**Comment by iandoug** [comment]
Hi Caleb

Should you not be sleeping? :-)

Sile works ... I was testing it yesterday but gave up because I actually wanted it to generate invoices and certificates, but it needs full Table support for that. I know that is not typical use case but trying to bypass potential issues with PHP external HTML->PDF package.

Am on Gentoo. 

Installed as root because that was necessary for it to install all the Lua deps. I did try as myself first.

Cheers, Ian

**Comment by Omikhleia** [comment]
> I actually wanted it to generate invoices and certificates, but it needs full Table support for that.

As a side note: An interesting use case for tables. You could try to see if those proposed in PR #1344 fit your needs, in terms of flexibility and options.

**Comment by iandoug** [comment]
@Omikhleia 

Trying to improve attached.

I actually looked at that yesterday. Then shook my head at Type Issues in Lua. One day all languages will be strictly typed and there will be less bugs.
Was planning on waiting for that to be integrated.

In the mean time I got brave and took another look at Tex ... installed assorted packages. Had Kile before but see it has not been updated in years and did not want to jump through hoops to use .otf / .ttf fonts. 

First attempt at tables worked (eventually), but sticking long text in cell produced same issue as this: "ptable is a possible solution for more advanced tables that support line-breaking in cells."

So will need to do it the hard way and learn more Tex to find the incantation that handles it. I assume Donald did handle those cases :-)

While I wait for Sile to continue maturing :-)

Thanks, Ian.
![2022-04-03_11-31](https://user-images.githubusercontent.com/12673886/161421435-aa5e3f0a-19be-4f77-aa70-b9bb92d94fe9.png)

**Comment by Omikhleia** [comment]
> (...) Then shook my head at Type Issues in Lua.

It's not clear to me what you are trying to do (and why or where Lua enters the equation here, vs. SILE-language)

> First attempt at tables worked (eventually), but sticking long text in cell produced same issue as this: "ptable is a possible solution for more advanced tables that support line-breaking in cells."

As far as I can tell, your above image does seem to fit the possibilities of the proposed "ptable" (i.e. I don't see anything here that would not be achievable with it)

> (...) learn more Tex to find the incantation that handles it. I assume Donald did handle those cases :-)

The tables in Knuth's (plain-)TeX are pretty low-level... There's some good reasons for LaTeX and Co to have so many different packages on top of that for more advanced tables: LaTeX's own "table" environment obviously, but also extra packages such as "booktabs", "ctable", etc. Tables are tough. Near the end of #45 (and also in the tests provided in the above-mentioned PR), I tried to ensure some complex cases from the latter could be reproduced with my "ptable". There are also some examples of code-generated tables.

(But we are lingering on being out of topic regarding the initial issue opened here, so I'll suggest to have this discussion elsewhere)

**Comment by iandoug** [comment]
> (and why or where Lua enters the equation here, vs. SILE-language)

Caleb's comment in your pull request. 
https://github.com/sile-typesetter/sile/pull/1344#issuecomment-1070944127

Thanks for your help.

**Comment by Omikhleia** [comment]
> > (and why or where Lua enters the equation here, vs. SILE-language)
> 
> Caleb's comment in your pull request. [#1344 (comment)](https://github.com/sile-typesetter/sile/pull/1344#issuecomment-1070944127)

Oh, maybe I wasn't clear in the subsequent comments, but that was fixed and the PR updated (twice) after that comment.

**Comment by iandoug** [comment]
Hi

mlocate ran overnight so this morning I could find where the file is.
/usr/local/share/sile/packagemanager/fontproof 

Owned by root so had to chmod.

Got fontproof to run, issues moved to that repo.

Thanks, Ian

**Comment by alerque** [comment]
Just a side note on Gentoo, I am officially a contributor now. To test the waters for a possible SILE package I [submitted some minor fixes](https://github.com/gentoo/gentoo/pull/24879) to Lua packaging and the process was quite painlessly accepted in a matter of hours. Following the pattern for Lua packages shouldn't be too hard, and once we have all the dependencies adding a SILE build shouldn't be a hard step either since it is straight autotools defaults compliant at this point.

**Comment by iandoug** [comment]
@alerque 
Cool :-)

There's also that Libertinus package... :-)

Thanks, Ian

**Comment by iandoug** [comment]
> The tables in Knuth's (plain-)TeX are pretty low-level... There's some good reasons for LaTeX and Co to have so many different packages on top of that for more advanced tables: LaTeX's own "table" environment obviously, but also extra packages such as "booktabs", "ctable", etc. Tables are tough. Near the end of #45 (and also in the tests provided in the above-mentioned PR), I tried to ensure some complex cases from the latter could be reproduced with my "ptable". There are also some examples of code-generated tables.

Comparison. Creating PDF with PHP.

https://peterdev.pl/picking-a-php-tool-to-generate-pdfs/

**Comment by Omikhleia** [comment]
> Comparison. Creating PDF with PHP.
> https://peterdev.pl/picking-a-php-tool-to-generate-pdfs/

It's kind of comparing apples and oranges, no? The above article mostly concludes on solutions that convert HTML/CSS to PDF, using a browser-based solution.

**Comment by iandoug** [comment]
I shared because of the samples which illustrate how different libraries struggle with doing table-based layouts. 

I've decided to take a look at WeasyPrint, even though I don't have much experience with Python, and it is slow, but it seems to be the most mature.

Don't want to run Chrome headless on a server.

**Comment by alerque** [comment]
Different workflows and use cases of course call for different tools, SILE isn't the best at everything!

[WeasyPrint](https://weasyprint.org) is quite nice for some situations. Also worth mentioning in this space are [speedata](https://www.speedata.de/en/), [paged.js](https://pagedjs.org), and [prince](https://www.princexml.com).

**Comment by iandoug** [comment]
@alerque 
Thanks ... need to make time to play with these.

**Comment by alerque** [comment]
[This comment on another issue](https://github.com/sile-typesetter/sile/issues/324#issuecomment-1169873737) is also relevant to this issue:

> As of #1450 which will be released in v0.14.0, SILE now supports packages installed via LuaRocks. This is documented in the manual and discussed in related PRs, but here is a brief overview re this issue:
> 
> * SILE has always scanned the default Lua `package.path` for packages when asked to load one in addition to it's own installation directory. Now it also appends an extra `/sile` to the default Lua paths before scanning them. That means `SILE.require(packages.foo)` will look in all the usual Lua places for things that are SILE specific packages (e.g. `/usr/share/lua/5.4/sile/?/init.lua`).
> * Package / class authors can create, publish, and install packages with the usual `luaarocks` tooling just by setting the module path in the rockspec to have SILE as the top level directory. This also allows 3rd party addons to override almost any core parts of SILE.
> * This enabled at least 4 possible installation locations:
>   1. System default (root usually required). These will automatically be used by all invocations of SILE.
>   2. User directory installations (`luarocks --local install`). Use in SILE will require exporting the paths fires, e.g. with `eval $(luarocks --local path)`.
>   3. Project local installations (`luarocks --tree ./ install`). Will work automatically when rendering documents in that directory.
>   4. Any custom library location(s) you choose, use `--tree` with the path to each location you want to load. Requires adding to the standard Lua path like option №2 before use.
> 
> Rocks can be published on luarocks.org, but they can also be in private repositories with the standard tooling. In most cases a LuaRocks enabled SILE addon repository could also be added as a submodule to a SILE project or just extracted to the same directory for use without having to even touch LuaRocks.

**Comment by alerque** [comment, marked as answer]
Skimming through the above, I *think* everything has been either addressed or been redirected to more specific issues. If I missed anything feel free to ask.
